src/trace_ng/log.zig

```diff
@@ -0,0 +1,672 @@
+const std = @import("std");
+const Level = @import("level.zig").Level;
+const logfmt = @import("logfmt.zig");
+const sig = @import("../sig.zig");
+// TODO Improve import.
+const Channel = @import("../sync/channel.zig").Channel;
+const testing = std.testing;
+const Allocator = std.mem.Allocator;
+const AtomicBool = std.atomic.Value(bool);
+const RecycleFBA = sig.utils.allocators.RecycleFBA;
+
+pub const Config = struct {
+    max_level: Level = Level.debug,
+    allocator: std.mem.Allocator,
+    /// Maximum memory that logger can use.
+    max_buffer: u64,
+    exit_sig: *std.atomic.Value(bool),
+    kind: LogKind = LogKind.standard,
+};
+
+const INITIAL_LOG_CHANNEL_SIZE: usize = 1024;
+
+const LogKind = enum {
+    standard,
+    testing,
+    noop,
+};
+
+const UnscopedLogger = StandardLogger(null);
+pub fn StandardLogger(comptime scope: ?type) type {
+    const StanardErrLogger = struct {
```

Comment:
> For more flexibility (and so people like me can understand it more easily) you could define these types outside the StandardLogger function. That way they can be directly initialized in other scopes. And it decouples the interface from the implementations. It keeps the interface simpler.

I remember trying this but ran into a snag because they need to capture the `comptime scope` scope. hence they need to be within the type function. 

I'll cycle back to this after processing the other comments

Reply:
Just a note: easy way to get around this would be to simply parameterize the implementations with the scope parameter as well.